Content.Shared/Administration/SharedNetworkResourceManager.cs

```diff
@@ -0,0 +1,83 @@
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using Robust.Shared.ContentPack;
+using Robust.Shared.Network;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.Administration;
+
+/// <summary>
+///     Manager that allows resources to be added at runtime by admins.
+///     They will be sent to all clients automatically.
+/// </summary>
+public abstract class SharedNetworkResourceManager : IContentRoot
+{
+    [Dependency] private readonly INetManager _netManager = default!;
+    [Dependency] protected readonly IResourceManager ResourceManager = default!;
+
+    public const double BytesToMegabytes = 0.000001d;
+
+    /// <summary>
+    ///     The prefix for any and all downloaded network resources.
+    /// </summary>
+    private static readonly ResourcePath Prefix = ResourcePath.Root / "Uploaded";
+
+    [ViewVariables]
+    protected readonly Dictionary<ResourcePath, byte[]> Files = new();
+
+    public virtual void Initialize()
+    {
+        _netManager.RegisterNetMessage<NetworkResourceUploadMessage>(ResourceUploadMsg);
+
+        // Add ourselves as a content root.
+        ResourceManager.AddRoot(Prefix, this);
+    }
+
+    protected abstract void ResourceUploadMsg(NetworkResourceUploadMessage msg);
+
+    public bool TryGetFile(ResourcePath relPath, [NotNullWhen(true)] out Stream? stream)
+    {
+        byte[]? data;
+
+        lock(Files)
```

Comment:
The locking strategy here should be a `ReaderWriterLockSlim`, not a `lock` mutex.

Reply:
Done. I hope it's correct now

Reply:
Also this requires a new entry in the sandbox whitelist, but I'm gonna move the impl of `IContentRoot` to engine so not a big deal for now